wallet/src/account/mod.rs

```diff
@@ -0,0 +1,407 @@
+                // inputs and utxos have the same length
+                let destination = Self::get_tx_output_destination(utxo).ok_or_else(|| {
+                    WalletError::UnsupportedTransactionOutput(Box::new(utxo.clone()))
+                })?;
+
+                if *destination == Destination::AnyoneCanSpend {
+                    Ok(InputWitness::NoSignature(None))
+                } else {
+                    let private_key =
+                        self.key_chain.get_private_key_for_destination(destination)?.private_key();
+
+                    let sighash_type =
+                        SigHashType::try_from(SigHashType::ALL).expect("Should not fail");
+
+                    StandardInputSignature::produce_uniparty_signature_for_input(
+                        &private_key,
+                        sighash_type,
+                        destination.clone(),
+                        &tx,
+                        &input_utxos,
+                        i,
+                    )
+                    .map(InputWitness::Standard)
+                    .map_err(WalletError::TransactionSig)
+                }
+            })
+            .collect::<Result<Vec<InputWitness>, _>>()?;
+
+        let tx = req.get_signed_transaction(witnesses)?;
+
+        Ok(tx)
+    }
+
+    pub fn account_index(&self) -> U31 {
+        self.key_chain.account_index()
+    }
+
+    /// Get the id of this account
+    pub fn get_account_id(&self) -> AccountId {
+        self.key_chain.get_account_id()
+    }
+
+    /// Get a new address that hasn't been used before
+    pub fn get_new_address<B: Backend>(
+        &mut self,
+        db_tx: &mut StoreTxRw<B>,
+        purpose: KeyPurpose,
+    ) -> WalletResult<Address> {
+        Ok(self.key_chain.issue_address(db_tx, purpose)?)
+    }
+
+    fn add_utxo_if_own<B: Backend>(
+        &mut self,
+        db_tx: &mut StoreTxRw<B>,
+        output: &TxOutput,
+        outpoint: OutPoint,
+        utxo_source: utxo::UtxoSource,
+    ) -> WalletResult<()> {
+        if self.is_available_for_spending(output) && self.is_mine_or_watched(output) {
+            let utxo = Utxo::new(output.clone(), utxo_source);
+            let account_utxo_id = AccountOutPointId::new(self.get_account_id(), outpoint);
+            db_tx.set_utxo(&account_utxo_id, utxo)?;
+            self.mark_output_as_used(db_tx, output)?;
+        }
+        Ok(())
+    }
+
+    fn del_utxo_if_own<B: Backend>(
+        &self,
+        db_tx: &mut StoreTxRw<B>,
+        outpoint: &OutPoint,
+    ) -> WalletResult<()> {
+        let account_utxo_id = AccountOutPointId::new(self.get_account_id(), outpoint.clone());
+        db_tx.del_utxo(&account_utxo_id)?;
+        Ok(())
+    }
+
+    fn is_available_for_spending(&self, _txo: &TxOutput) -> bool {
+        // TODO implement
+        true
+    }
+
+    fn get_tx_output_destination(txo: &TxOutput) -> Option<&Destination> {
```

Comment:
Proper destination getter is implemented in `chainstate/tx-verifier/src/transaction_verifier/signature_destination_getter.rs`
I'd place a todo here and return something like `UnsupportedTransactionOutput`

Reply:
To clarify a bit: I suggest returning error for the match arms that require accounting info, not the whole function

Reply:
Sorry, I don't quite understand what you mean. I think this code will change and we could leave it as is for now.

Reply:
Retrieving the destination from an output is not as simple as it looks. It depends on the task at hand. It's simple for normal Transfer outputs, but when staking is involved, for example, it becomes more complex. I think Gosha means that there's a facility now for how to do that (I created that a few days ago). So, you can use that in order to handle this problem correctly, whether through using the already-provided implementations, or adding your own for your own wallet-related purposes.

Reply:
I think it's too early to spend a lot of time on that. Once we support staking/decommissioning, we could think about that. I added a few TODOs.